File: frontend/services/llm.ts
import { supabase } from './supabase';

// Type declaration for process.env
declare const process: {
  env: {
    NEXT_PUBLIC_API_URL?: string;
    [key: string]: string | undefined;
  };
};

// API Base URL
// For client-side code, use the environment variable or localhost
// The environment variable is set in docker-compose.yml
const API_URL = typeof window !== 'undefined'
  ? (process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000')
  : 'http://backend:8000'; // When running server-side in Docker, use the service name

// Type definitions
export interface TextGenerationRequest {
  prompt: string;
  model?: string;
  max_tokens?: number;
  temperature?: number;
  provider?: 'openai' | 'anthropic';
}

export interface TextGenerationResponse {
  text: string;
  model: string;
  usage: {
    prompt_tokens: number;
    completion_tokens: number | null;
    total_tokens: number;
  };
}

export interface EmbeddingRequest {
  text: string;
  model?: string;
  provider?: 'openai' | 'anthropic';
}

export interface EmbeddingResponse {
  embedding: number[];
  model: string;
  usage: {
    prompt_tokens: number;
    completion_tokens: number | null;
    total_tokens: number;
  };
}

// Check if we're in demo mode
const demoModeEnv = process.env.NEXT_PUBLIC_DEMO_MODE || '';
const isDemoMode = demoModeEnv.toLowerCase() === 'true' || 
                   !process.env.NEXT_PUBLIC_SUPABASE_URL || 
                   process.env.NEXT_PUBLIC_SUPABASE_URL === '';

// Helper to get authentication token
async function getAuthToken() {
  if (isDemoMode) {
    return 'demo-token'; // Mock token for demo mode
  }
  
  const { data: { session } } = await supabase.auth.getSession();
  return session?.access_token;
}

// Helper to make authenticated API requests
async function apiRequest(endpoint: string, method: string, body?: any) {
  // In demo mode, use the demo endpoint instead
  if (isDemoMode && endpoint === '/api/llm/generate') {
    try {
      console.log(`Making demo request to ${API_URL}/api/llm/demo`);
      
      const response = await fetch(`${API_URL}/api/llm/demo`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        throw new Error(`Demo API request failed with status ${response.status}`);
      }

      const data = await response.json();
      
      // Transform demo response to match expected format
      return {
        text: data.content,
        model: data.model,
        usage: {
          prompt_tokens: Math.floor(Math.random() * 50) + 10,
          completion_tokens: Math.floor(Math.random() * 100) + 20,
          total_tokens: Math.floor(Math.random() * 150) + 30
        }
      };
    } catch (error) {
      console.error('Demo API request error:', error);
      throw error;
    }
  }

  const token = await getAuthToken();

  if (!token) {
    throw new Error('Authentication required');
  }

  try {
    console.log(`Making request to ${API_URL}${endpoint}`);

    const response = await fetch(`${API_URL}${endpoint}`, {
      method,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/json',
      },
      mode: 'cors',
      body: body ? JSON.stringify(body) : undefined,
    });

    console.log(`Response status: ${response.status}`);

    if (!response.ok) {
      let errorMessage = `API request failed with status ${response.status}`;
      try {
        const errorData = await response.json();
        errorMessage = errorData.detail || errorMessage;
      } catch (e) {
        // If we can't parse the error as JSON, just use the status message
      }
      throw new Error(errorMessage);
    }

    return response.json();
  } catch (error) {
    console.error('API request error:', error);
    throw error;
  }
}

// LLM API functions
export async function generateText(request: TextGenerationRequest): Promise<TextGenerationResponse> {
  return apiRequest('/api/llm/generate', 'POST', request);
}

export async function createEmbedding(request: EmbeddingRequest): Promise<EmbeddingResponse> {
  return apiRequest('/api/llm/embedding', 'POST', request);
}

File: backend/app/api/endpoints/llm.py
from fastapi import APIRouter, Depends, HTTPException, status, Request
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
import logging
from typing import Optional

from app.services.llm.llm_service import LLMService, get_llm_service
from app.services.llm.embedding_service import EmbeddingService, get_embedding_service
from app.models.llm import TextGenerationRequest, TextGenerationResponse, EmbeddingRequest, EmbeddingResponse
from app.services.supabase.auth import SupabaseAuthService, get_auth_service
from app.core.config import settings
from app.core.demo import demo_service

router = APIRouter()
security = HTTPBearer()  # Make authentication required
logger = logging.getLogger(__name__)

# Import rate limiter
from app.core.rate_limiter import limiter


@router.post("/demo", response_model=dict)
@limiter.limit("10/minute")
async def demo_generate(request_data: dict, request: Request):
    """Demo endpoint that works without authentication or API keys."""
    try:
        prompt = request_data.get("prompt", "Hello from the Vibe Stack!")
        model = request_data.get("model", "demo")
        
        # Check if we're in demo mode
        if demo_service.is_demo_mode(settings.SUPABASE_URL, settings.OPENAI_API_KEY, settings.DEMO_MODE):
            response = demo_service.mock_llm_response(prompt, model)
            return {
                "content": response["content"],
                "model": response["model"],
                "demo": True,
                "message": "🌟 Demo mode active! Add your API keys to use real AI services."
            }
        
        # If we have real API keys, use the real service
        return {
            "content": "Real API integration is available. Use the /generate endpoint with authentication.",
            "demo": False
        }
        
    except Exception as e:
        logger.error(f"Demo generation error: {str(e)}")
        return {
            "content": "Demo mode is working! This shows how AI integration works in the Vibe Stack.",
            "model": "demo",
            "demo": True,
            "error": str(e)
        }


@router.post("/generate", response_model=TextGenerationResponse)
@limiter.limit("30/minute")
async def generate_text(
    request: TextGenerationRequest,
    credentials: HTTPAuthorizationCredentials = Depends(security),
    auth_service: SupabaseAuthService = Depends(get_auth_service),
    # We will override this service based on the request provider
    llm_service: LLMService = Depends(lambda: get_llm_service("openai")),
):
    """Generate text using the specified LLM model."""
    try:
        # Log request details for debugging
        logger.info(f"Sameer Received text generation request with model: {request}, provider: {credentials}")

        # Validate user authentication
        try:
            user = await auth_service.get_user(credentials.credentials)
            logger.info(f"User authenticated: {user.email if hasattr(user, 'email') else 'Unknown user'}")
        except Exception as auth_error:
            logger.error(f"Authentication error: {str(auth_error)}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Authentication failed: {str(auth_error)}",
                headers={"WWW-Authenticate": "Bearer"},
            )

        # Get the right LLM service based on provider
        try:
            if request.provider:
                llm_service = get_llm_service(request.provider)
                logger.info(f"Using LLM provider: {request.provider}")
        except ValueError as provider_error:
            logger.error(f"Provider error: {str(provider_error)}")
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(provider_error))

        # Generate text with the LLM service
        try:
            logger.info(f"Generating text with prompt: {request.prompt[:50]}...")

            # Check if API keys are configured
            if request.provider == "openai" and not settings.OPENAI_API_KEY:
                raise ValueError("OpenAI API key not configured. Please set the OPENAI_API_KEY environment variable.")
            elif request.provider == "anthropic" and not settings.ANTHROPIC_API_KEY:
                raise ValueError("Anthropic API key not configured. Please set the ANTHROPIC_API_KEY environment variable.")

            response = await llm_service.generate_text(
                prompt=request.prompt, model=request.model, max_tokens=request.max_tokens, temperature=request.temperature
            )
            logger.info(f"Text generation successful, response length: {len(response.text)}")
            return TextGenerationResponse(text=response.text, model=response.model, usage=response.usage)
        except Exception as generation_error:
            logger.error(f"Text generation error: {str(generation_error)}", exc_info=True)
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Text generation failed: {str(generation_error)}")

    except HTTPException:
        # Re-raise HTTP exceptions so they maintain their status codes
        raise
    except Exception as e:
        logger.error(f"Unexpected error: {str(e)}", exc_info=True)
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Unexpected error: {str(e)}")


@router.post("/embedding", response_model=EmbeddingResponse)
async def create_embedding(
    request: EmbeddingRequest,
    credentials: HTTPAuthorizationCredentials = Depends(security),
    auth_service: SupabaseAuthService = Depends(get_auth_service),
    embedding_service: EmbeddingService = Depends(get_embedding_service),
):
    """Create an embedding vector for the provided text."""
    try:
        # Validate user authentication
        try:
            user = await auth_service.get_user(credentials.credentials)
            logger.info(f"User authenticated: {user.email if hasattr(user, 'email') else 'Unknown user'}")
        except Exception as auth_error:
            logger.error(f"Authentication error: {str(auth_error)}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Authentication failed: {str(auth_error)}",
                headers={"WWW-Authenticate": "Bearer"},
            )

        # Generate embedding with the embedding service
        embedding = await embedding_service.create_embedding(text=request.text, model=request.model)

        return EmbeddingResponse(embedding=embedding.embedding, model=embedding.model, usage=embedding.usage)
    except Exception as e:
        logger.error(f"Embedding creation failed: {str(e)}", exc_info=True)
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Embedding creation failed: {str(e)}")
